### PR TITLE
Calphy implementation in pyiron

### PR DESCRIPTION
**Pull request by srmnitc**
This pull request implements calphy wrapper for pyiron. The calphy wrapper is implemented in `pyiron_atomistics/calphy`.

Calphy wrapper adds the possibility for fast free energy calculations for interatomic potentials.

### CODE REVIEW

**Comment by srmnitc** [comment]
@pmrv I have an issue in which I would like some help:
When loading a finished job, I need to explicitely call `from_hdf`. See [here](https://github.com/srmnitc/pyiron_atomistics/blob/master/tests/calphy/test_base.py#L74). Could you maybe help me with this issue?

**Comment by pmrv** [comment]
> @pmrv I have an issue in which I would like some help: When loading a finished job, I need to explicitely call `from_hdf`. See [here](https://github.com/srmnitc/pyiron_atomistics/blob/master/tests/calphy/test_base.py#L74). Could you maybe help me with this issue?

Sure, you're not supposed to instantiate a job like this directly.  So instead of
```python
cls.output_job = Calphy(
    project=ProjectHDFio(project=cls.output_project, file_name="tm_fcc"),
    job_name="tm_fcc",
)
```
you want
```python
cls.output_job = cls.output_project.load("tm_fcc")
```
If that doesn't work the project is setup wrongly.

**Comment by pmrv** [comment]
Ah, I see you try to load some prre-run job.  What you probably want to do is more like [this](https://github.com/pyiron/pyiron_atomistics/blob/ce840e5434de3c0e78d575be10b4e37f9abf6826/tests/atomistics/job/test_transform_trajectory.py#L54).

Prepare the job in a project on the cluster.  `Project.pack` them stick that archive in `tests/static` and then `Project.unpack` them in the `setUp` method of your tests.

File: pyiron_atomistics/calphy/calphy.py
Comment (pmrv) [inline pyiron_atomistics/calphy/calphy.py]: `GenericJob` does this automatically now.
```suggestion
```

File: pyiron_atomistics/calphy/calphy.py
Comment (pmrv) [inline pyiron_atomistics/calphy/calphy.py]: I know Lammps and Vasp have this pattern in our code, but imo it's bad form.  This class doesn't implement anything on top of `CalphyBase` and (I guess) we do not plan on having multiple classes derive from it, so best to keep everything in one spot.

File: tests/calphy/test_base.py
Comment (pmrv) [inline tests/calphy/test_base.py]: See my top-level comments regarding `pack`/`unpack`.

File: tests/calphy/test_base.py
Comment (pmrv) [inline tests/calphy/test_base.py]:
```suggestion
    @classmethod
    def tearDownClass(cls):
        cls.execution_path = os.path.dirname(os.path.abspath(__file__))
        project = Project(os.path.join(cls.execution_path, "test_calphy"))
        project.remove_jobs_silently(recursive=True)
        project.remove(enable=True)
        state.update()
```
You'll want to derive from `pyiron_base._tests.TestWithProject`.

File: tests/calphy/test_base.py
Comment (pmrv) [inline tests/calphy/test_base.py:50]: Just a general comment regarding unit tests: I personally strongly prefer short docstrings and passing `msg=...` to `self.assert...`.  No need to go overboard with this, but if I come back in half a year I should be able to understand the rationale for the test and what invariant it is testing without reading all the involved code.

File: pyiron_atomistics/calphy/base.py
Comment (pmrv) [inline pyiron_atomistics/calphy/base.py]:
```suggestion
        self.__name__ = "CalphyJob"
```
No need, done by `GenericJob` automatically.

File: pyiron_atomistics/calphy/base.py
Comment (pmrv) [inline pyiron_atomistics/calphy/base.py]: If this not something (Notebook-) users are expected to call, rename with underscore -> `_prepare_pair_styles`.

File: pyiron_atomistics/calphy/base.py
Comment (pmrv) [inline pyiron_atomistics/calphy/base.py]:
```suggestion
            list of str: list of interatomic potentials
```

File: pyiron_atomistics/calphy/base.py
Comment (pmrv) [inline pyiron_atomistics/calphy/base.py]:
```suggestion
    def get_potentials(self) -> List[str]:
```

File: pyiron_atomistics/calphy/base.py
Comment (pmrv) [inline pyiron_atomistics/calphy/base.py]:
```suggestion
from typing import Union, List
```

File: pyiron_atomistics/calphy/base.py
Comment (pmrv) [inline pyiron_atomistics/calphy/base.py]: I guess this is in analogy to `AtomisticGenericJob`, but I think it's a bit of a false friend. I'd either rename in plural or make it operate explicitly only on the initial potential.  Otherwise users can be in a situation where they (by above analogy) expect a string back, but get a list.

File: pyiron_atomistics/calphy/base.py
Comment (pmrv) [inline pyiron_atomistics/calphy/base.py]:
```suggestion
        
        Args:
            None
        
        Returns:
            None
```

**Review by pmrv — CHANGES_REQUESTED** [review]
Looks generally pretty good, but I left some nits.  Especially for typing I like it to be as explicit as possible, e.g. using `List[str]` instead of just `list`. 

Only bigger concern is the naming of the `calc_*` methods.  Are all of them intended as entry points for users? If so, everythings good, if not I can see there being a bit of confusion between `calc_mod_*` and `calc_free_energy`, so a bit more explanation would be nice.

**Comment by srmnitc** [comment]
> Looks generally pretty good, but I left some nits. Especially for typing I like it to be as explicit as possible, e.g. using `List[str]` instead of just `list`.
> 
> Only bigger concern is the naming of the `calc_*` methods. Are all of them intended as entry points for users? If so, everythings good, if not I can see there being a bit of confusion between `calc_mod_*` and `calc_free_energy`, so a bit more explanation would be nice.

I updated the docstrings to clarify the methods.

File: pyiron_atomistics/calphy/calphy.py
Comment (srmnitc) [reply pyiron_atomistics/calphy/calphy.py]: I am not sure I follow fully the changes. In principle, I could remove `calphy.py` altogether. Would this be the best idea?

Comment (pmrv) [reply pyiron_atomistics/calphy/calphy.py]: Yes, that's want I recommend.  Remove `Calphy` and rename `CalphyBase` to `Calphy`.  I'd probably then also rename the file from `base.py` to `job.py` or something similar, because words like `base`, `generic`, etc. have very little information content.

File: tests/calphy/test_base.py
Comment (srmnitc) [reply tests/calphy/test_base.py]: changed tests

File: pyiron_atomistics/calphy/base.py
Comment (srmnitc) [reply pyiron_atomistics/calphy/base.py]: Done.

File: pyiron_atomistics/calphy/calphy.py
Comment (srmnitc) [reply pyiron_atomistics/calphy/calphy.py]: I renamed the files and removed base

File: pyiron_atomistics/calphy/base.py
Comment (srmnitc) [reply pyiron_atomistics/calphy/base.py]: I would like to keep the property as `potential`. The logic behind this is that if you set two potentials, the mode is `alchemy` automatically. I updated the property to return a string if only one potential was set. That should remove the confusion.

File: pyiron_atomistics/calphy/job.py
Comment (pmrv) [inline pyiron_atomistics/calphy/job.py]:
```suggestion
```
Flagged by codacy.

File: pyiron_atomistics/calphy/job.py
Comment (pmrv) [inline pyiron_atomistics/calphy/job.py]:
```suggestion
        super().__init__(project, job_name)
```

File: tests/calphy/test_base.py
Comment (pmrv) [inline tests/calphy/test_base.py]:
```suggestion
import unittest
```
Also codacy.

File: tests/calphy/test_base.py
Comment (pmrv) [inline tests/calphy/test_base.py]: `cwd` is not defined, probably you want something like `os.path.basename(__FILE__)`.

**Review by pmrv — APPROVED** [review]
I highlighted some typos and such from codacy, but otherwise looks good to me!

I'm surprised that the tests appear to run, because there's one undefined variable, please double check locally that everything is good and have a look at the coverage before merging.

File: tests/calphy/test_base.py
Comment (srmnitc) [reply tests/calphy/test_base.py]: changed to `os.getcwd()`

**Comment by srmnitc** [comment]
> I highlighted some typos and such from codacy, but otherwise looks good to me!
> 
> I'm surprised that the tests appear to run, because there's one undefined variable, please double check locally that everything is good and have a look at the coverage before merging.

I am also surprised about this. However, I have now checked the tests locally too and everything seems fine. I will go ahead and merge then. Thank you for your review and comments!

**Comment by pmrv** [comment]
> > I highlighted some typos and such from codacy, but otherwise looks good to me!
> > I'm surprised that the tests appear to run, because there's one undefined variable, please double check locally that everything is good and have a look at the coverage before merging.
> 
> I am also surprised about this. However, I have now checked the tests locally too and everything seems fine. I will go ahead and merge then. Thank you for your review and comments!

Ah, I found the issue.  You'll need to place an empty `__init__.py` into `tests/calphy`, otherwise `unittests` will not recognize it as a module and won't run the tests.

You can force it to run them by doing
`python -m unittest tests/calphy/test_calphy.py`.

**Comment by srmnitc** [comment]
> > > I highlighted some typos and such from codacy, but otherwise looks good to me!
> > > I'm surprised that the tests appear to run, because there's one undefined variable, please double check locally that everything is good and have a look at the coverage before merging.
> > 
> > 
> > I am also surprised about this. However, I have now checked the tests locally too and everything seems fine. I will go ahead and merge then. Thank you for your review and comments!
> 
> Ah, I found the issue. You'll need to place an empty `__init__.py` into `tests/calphy`, otherwise `unittests` will not recognize it as a module and won't run the tests.
> 
> You can force it to run them by doing `python -m unittest tests/calphy/test_calphy.py`.

Ah, it does not discover the tests; I have been running the file directly locally, which is why I did not have issues. Interesting! thanks!

**Comment by srmnitc** [comment]
Ready to merge, but kept on hold due to [the license discussion](https://github.com/orgs/pyiron/teams/pyiron/discussions/116?from_comment=2#discussion-116-comment-2)

**Comment by srmnitc** [comment]
> Ready to merge, but kept on hold due to [the license discussion](https://github.com/orgs/pyiron/teams/pyiron/discussions/116?from_comment=2#discussion-116-comment-2)

Released Calphy API under BSD license, removing the issues that prevent this merge.

**Pull request merged**
